### PR TITLE
[ASCII-2410] remove usage of global tagger 

### PR DESCRIPTION
**Pull request by GustavoCaso**
### What does this PR do?

This PR remove the global tagger accessor from he Agent. 

### Motivation

We need this for a couple reasons:
- Reduce usage of globals across the codebase
- We want to migrate the tagger component to the latest component framework file structure 
   - **That would allow us to fully decouple the remote tagger from their other implementations. Otel, and trace agent and any other process that uses the remote tagger would benefit from this, as they no longer need to import all the tagger dependencies.**

All changes involve passing the tagger component to places that previously was using the global tagger accessor directly. 

A few notes on the changes:
- In the tests that we not using the tagger component before and did not needed the tagger at all for working I uses the `nooptagger`. That reduces the amount of changes for those tests
- We couldn't scape copmpletely the use of globals. I had to overcome some limitations with the exiting depency graph for components and initialize the container provider as part of Fx start process 😢 
- For the autodiscovery listenerts I create a new struct `ServiceListernerOptions`. That way if in the future we need to add a new argument to the listener init process we do not have to change many function definitions

##### Python integrations

We need to verify that python integrations can still query the tagger normally.

<details>
<summary>Instructions for QA python checks </summary>
<br>
Deploy the agent with helm.

Do the following steps to deploy a custom python check onto the agent:

1. Check the running pods:
```
kubectl get pods
NAME                                         READY   STATUS    RESTARTS      AGE
datadog-agent-cluster-agent-cfc87c6c-d26pg   1/1     Running   0             107m
datadog-agent-hsstl                          3/3     Running   4 (73m ago)   107m
```

2. Check the content of the tagger, and pick any tag entity id for testing:
```
kubectl exec datadog-agent-hsstl -- agent tagger-list
....
=== Entity kubernetes_pod_uid://b97232df7635c1e88c7bd9309b8e78ae ===
== Source workloadmeta-kubernetes_pod =
=Tags: [kube_namespace:kube-system kube_priority_class:system-node-critical kube_qos:Burstable pod_name:kube-apiserver-kind-control-plane pod_phase:running]
===
....
```

3. Exec into the agent container: `kubectl exec -it datadog-agent-hsstl -c agent -- bin/bash`
4. Add a file called `custom_checkvalue.yaml` under `etc/datadog-agent/conf.d` with the following content:
```
init_config:
instances:
  [{}]
```
5. Add the check code in a file called `custom_checkvalue.py` under `etc/datadog-agent/checks.d` with the following content:

```
from datadog_checks.base.utils.tagging import tagger
from checks import AgentCheck
class HelloCheck(AgentCheck):
  def check(self, instance):
    tags = tagger.tag('kubernetes_pod_uid://b97232df7635c1e88c7bd9309b8e78ae', tagger.HIGH)
    print("obtained tags: ", tags)
    self.gauge('hello.world', 1, tags)
```

ATTENTION: you should use the entity id that you go from the tagger-list command in step 2.

6. Restart the agent by running: `agent stop && agent start`

Wait for the agent to start, and run the custom check as follows:

```
 kubectl exec datadog-agent-hsstl -- agent check custom_checkvalue > output.txt

....
=== Series ===
[
  {
    "metric": "hello.world",
    "points": [
      [
        1728996360,
        1
      ]
    ],
    "tags": [
      "kube_namespace:kube-system",
      "kube_priority_class:system-node-critical",
      "kube_qos:Burstable",
      "pod_name:kube-apiserver-kind-control-plane",
      "pod_phase:running"
    ],
    "host": "kind-control-plane-adel-test-rtloader",
    "type": "gauge",
    "interval": 0,
    "source_type_name": "System"
  }
]
....
```

Note: you migh not get the tags on each run due to lag of initialisation of workloadmeta and tagger. However, as long as you are getting the tags at least once, you can consider the QA as done. You can also check the Datadog UI to validate that you can see the metric and aggregate it by tags:

image
</details>

### CODE REVIEW

**Review by kangyili — APPROVED** [review]

**Review by NouemanKHAL — APPROVED** [review]

**Review by dineshg13 — APPROVED** [review]

**Review by AlexandreYang — APPROVED** [review]

**Review by HadhemiDD — APPROVED** [review]

File: pkg/aggregator/aggregator.go [hunk 1]
```diff
@@ -261,10 +262,12 @@ type BufferedAggregator struct {
...
-	tlmContainerTagsEnabled bool                                         // Whether we should call the tagger to tag agent telemetry metrics
-	agentTags               func(types.TagCardinality) ([]string, error) // This function gets the agent tags from the tagger (defined as a struct field to ease testing)
-	globalTags              func(types.TagCardinality) ([]string, error) // This function gets global tags from the tagger when host tags are not available
-
+	tlmContainerTagsEnabled     bool                                         // Whether we should call the tagger to tag agent telemetry metrics
+	agentTags                   func(types.TagCardinality) ([]string, error) // This function gets the agent tags from the tagger (defined as a struct field to ease testing)
+	globalTags                  func(types.TagCardinality) ([]string, error) // This function gets global tags from the tagger when host tags are not available
+	enrichTags                  func(tagset.TagsAccumulator, taggertypes.OriginInfo)
+	checksCardinality           func() types.TagCardinality
+	tagger                      tagger.Component
```
Comment (vickenty) [inline pkg/aggregator/aggregator.go:270]: Maybe add a couple blank lines to avoid shifting types and comments everything so far right.

```suggestion
	tlmContainerTagsEnabled     bool                                         // Whether we should call the tagger to tag agent telemetry metrics

	agentTags                   func(types.TagCardinality) ([]string, error) // This function gets the agent tags from the tagger (defined as a struct field to ease testing)
	globalTags                  func(types.TagCardinality) ([]string, error) // This function gets global tags from the tagger when host tags are not available
	enrichTags                  func(tagset.TagsAccumulator, taggertypes.OriginInfo)
	checksCardinality           func() types.TagCardinality
	tagger                      tagger.Component

```

File: comp/core/tagger/component.go [hunk 2]
```diff
@@ -38,6 +38,7 @@ type Component interface {
 	Stop() error
 	ReplayTagger() ReplayTagger
 	GetTaggerTelemetryStore() *telemetry.Store
+	LegacyTag(entity string, cardinality types.TagCardinality) ([]string, error)
```
Comment (adel121) [inline comp/core/tagger/component.go:41]: Can we please add a comment here explaining why we still have this `LegacyTag` method? 

Basically something similar to [this](https://github.com/DataDog/datadog-agent/blob/8de7847264222e3558f7a06c1572b633b183f5fe/comp/core/tagger/global.go#L48-L51):

Maybe something like:

```
// LegacyTag has the same behaviour as the Tag method, but it receives the entity id as a string and parses it. 
// If possible, avoid using this function, and use the Tag method instead.
// This function exists in order not to break backward compatibility with rtloader and python
// integrations using the tagger
```

File: pkg/aggregator/aggregator.go [hunk 3]
```diff
@@ -336,6 +339,9 @@ func NewBufferedAggregator(s serializer.MetricSerializer, eventPlatformForwarder
 		tlmContainerTagsEnabled:     pkgconfigsetup.Datadog().GetBool("basic_telemetry_add_container_tags"),
 		agentTags:                   tagger.AgentTags,
 		globalTags:                  tagger.GlobalTags,
+		enrichTags:                  tagger.EnrichTags,
```
Comment (vickenty) [inline pkg/aggregator/aggregator.go:342]: Is there a reason to store the method in a field? `agentTags` and `globalTags` are overridden in tests, but not this one. I think it is a legacy pre-components pattern, no need to copy it if we're not using it.

**Review by ichinaski — APPROVED** [review]

File: pkg/aggregator/aggregator_test.go [hunk 4]
```diff
@@ -575,9 +581,12 @@ func TestTags(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			defer pkgconfigsetup.Datadog().SetWithoutSource("basic_telemetry_add_container_tags", nil)
-			pkgconfigsetup.Datadog().SetWithoutSource("basic_telemetry_add_container_tags", tt.tlmContainerTagsEnabled)
-			agg := NewBufferedAggregator(nil, nil, tt.hostname, time.Second)
+			mockConfig := configmock.New(t)
+			mockConfig.SetWithoutSource("basic_telemetry_add_container_tags", tt.tlmContainerTagsEnabled)
```
Comment (vickenty) [inline pkg/aggregator/aggregator_test.go:585]: `mockConfig` created and not used? Are these changes necessary for the tagger replacement? If not, would it be possible to do that in a separate PR?

File: comp/aggregator/demultiplexer/demultiplexerimpl/status_test.go [hunk 5]
```diff
@@ -53,12 +55,17 @@ func TestStatusOutPut(t *testing.T) {
 		}},
 	}
 
+	mockTagger := taggerimpl.SetupFakeTagger(t)
```
Comment (vickenty) [inline comp/aggregator/demultiplexer/demultiplexerimpl/status_test.go:58]: There is this, `noopimpl.NewTaggerClient()`, `nooptagger.Module()`, `fxutil.Test[tagger.Mock](b, taggerimpl.MockModule())`. Why for different ways to set up a tagger for tests?

**Comment by adel121** [comment]
This PR touches the rtloader part that translate calls to the tagger done by the python integrations.

I think it can be good to verify that the conversion is still working as expected. ([example QA](https://github.com/DataDog/datadog-agent/pull/30133))

File: pkg/aggregator/aggregator.go [hunk 6]
```diff
@@ -336,6 +339,9 @@ func NewBufferedAggregator(s serializer.MetricSerializer, eventPlatformForwarder
...
+		checksCardinality:           tagger.ChecksCardinality,
```
Comment (vickenty) [inline pkg/aggregator/aggregator.go:343]: Same

**Comment by GustavoCaso** [comment]
> I think it can be good to verify that the conversion is still working as expected.

@adel121 add the QA instructions

File: pkg/aggregator/aggregator.go (hunk 3 above)
Comment (GustavoCaso) [reply pkg/aggregator/aggregator.go:342]: Sounds good. I wasn't sure what pattern to follow. I will update the code

File: comp/aggregator/demultiplexer/demultiplexerimpl/status_test.go (hunk 5 above)
Comment (GustavoCaso) [reply comp/aggregator/demultiplexer/demultiplexerimpl/status_test.go:58]: The tagger component mock is a bit of a mess. I would cleanup up on a following PR

File: comp/core/tagger/component.go (hunk 2 above)
Comment (adel121) [reply comp/core/tagger/component.go:41]: If I am not mistaken, when we command+click on a tagger component instance from another component, it redirects us to the component interface. That's why I would love to this warning comment visible easily to programmers to avoid missing it.

**Review by clarkb7 — APPROVED** [review]

**Review by robertjli — APPROVED** [review]

File: comp/core/tagger/taggerimpl/local/tagger.go [hunk 7]
```diff
@@ -105,6 +106,20 @@ func (t *Tagger) Tag(entityID types.EntityID, cardinality types.TagCardinality)
 	return tags.Copy(), nil
 }
 
+// LegacyTag is an interface function that queries taggerclient singleton
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/local/tagger.go:109]: I am not sure if the comment here is still accurate.

After this change I think the tagger is no longer a global singleton, but rather a component.

Maybe it would be better to have the comment as:

```
// LegacyTag queries the tagger component for tags based on a string entity id.
```

WDYT?

File: comp/core/tagger/taggerimpl/remote/tagger.go [hunk 8]
```diff
@@ -212,6 +213,20 @@ func (t *Tagger) Tag(entityID types.EntityID, cardinality types.TagCardinality)
 	return []string{}, nil
 }
 
+// LegacyTag is an interface function that queries taggerclient singleton
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/remote/tagger.go:216]: Same comment as above

File: comp/core/tagger/taggerimpl/replay/tagger.go [hunk 9]
```diff
@@ -69,6 +70,20 @@ func (t *Tagger) Tag(entityID types.EntityID, cardinality types.TagCardinality)
 	return tags, nil
 }
 
+// LegacyTag is an interface function that queries taggerclient singleton
```
Comment (adel121) [inline comp/core/tagger/taggerimpl/replay/tagger.go:73]: Same comment as above.

Actually in order to avoid repeating this comment in that many places, maybe we could just put the comment as:

```
// LegacyTag implements the LegacyTag method of the Tagger component
```

And then we can put the explanatory comment on the interface method directly.

**Review by adel121 — COMMENTED** [review]
Added some comments.

**Review by adel121 — APPROVED** [review]
LGTM for #container-platform owned files.

**Review by duncanista — APPROVED** [review]
LGTM on the Serverless side!

File: pkg/process/checks/net.go [hunk 10]
```diff
@@ -122,7 +122,11 @@ func (c *ConnectionsCheck) Init(syscfg *SysProbeConfig, hostInfo *HostInfo, _ bo
 	c.processData.Register(c.serviceExtractor)
 
 	// LocalResolver is a singleton LocalResolver
-	c.localresolver = resolver.NewLocalResolver(proccontainers.GetSharedContainerProvider(c.wmeta), clock.New(), maxResolverAddrCacheSize, maxResolverPidCacheSize)
+	sharedContainerProvider, err := proccontainers.GetSharedContainerProvider()
```
Comment (hmahmood) [inline pkg/process/checks/net.go:125]: I don't think this would work for process-agent which this connections check is part of. Looks like `proccontainers.InitSharedContainerProvider` is only called from the agent and cluster-agent binaries. @daniel-taf can you chime in here? I am not sure with the new single agent binary thing if this is a valid concern.

File: pkg/aggregator/aggregator_test.go (hunk 4 above)
Comment (vickenty) [reply pkg/aggregator/aggregator_test.go:585]: I'm not sure I follow. `basic_telemetry_add_container_tags` is used by the aggregator, not the tagger, and is read from the `pkgconfigsetup`. If it is set on the mock, will the code under testing here see the correct value?

File: pkg/process/checks/net.go (hunk 10 above)
Comment (daniel-taf) [reply pkg/process/checks/net.go:125]: IIRC single agent binary is not moving forward (but I could be out of the loop). As it is, this would not work as you say, the process-agent binary should also call `proccontainers.InitSharedContainerProvider`. In fact, I would expect our e2e tests to fail due to this.

Comment (GustavoCaso) [reply pkg/process/checks/net.go:125]: Nice catch @hmahmood 

I added the call to  `proccontainers.InitSharedContainerProvider` as part of the process-agent boot process.

Comment (hmahmood) [reply pkg/process/checks/net.go:125]: Thanks, could you also remove the field `ConnectionsCheck.wmeta`; it is not used anymore.

**Review by Yumasi — APPROVED** [review]
Approved for USM files

File: cmd/agent/subcommands/jmx/command.go [hunk 11]
```diff
@@ -159,6 +160,11 @@ func Commands(globalParams *command.GlobalParams) []*cobra.Command {
 			taggerimpl.Module(),
 			autodiscoveryimpl.Module(),
 			agent.Bundle(jmxloggerimpl.NewCliParams(cliParams.logFile)),
+			// InitSharedContainerProvider must be called before the application starts so the workloadmeta collector can be initiailized correctly.
+			// Since the tagger depends on the workloadmeta collector, we can not make the tagger a dependency of workloadmeta as it would create a circular dependency.
```
Comment (hush-hush) [inline cmd/agent/subcommands/jmx/command.go:164]: Could we add a `TODO - component` here to track the cleanup work once we remove the cycle deps between workloadmeta and the tagger ?

File: cmd/agent/subcommands/flare/command.go [hunk 12]
```diff
@@ -126,8 +127,12 @@ func Commands(globalParams *command.GlobalParams) []*cobra.Command {
 					InitHelper: common.GetWorkloadmetaInit(),
 					NoInstance: !cliParams.forceLocal, //if forceLocal is true, we want to run workloadmeta
 				}),
-				taggerimpl.OptionalModule(),
-				autodiscoveryimpl.OptionalModule(), // if forceLocal is true, we will start autodiscovery (loadComponents) later
+				fx.Provide(tagger.NewTaggerParams),
+				taggerimpl.Module(),
```
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply cmd/agent/subcommands/flare/command.go:131]: I would say that belong to another PR. I rather avoid adding more changes to this PR. We can do them on a following PR

File: pkg/aggregator/aggregator_test.go (hunk 4 above)
Comment (GustavoCaso) [reply pkg/aggregator/aggregator_test.go:585]: Prior to this change, we were not using `fxutil.Test` in the tests. Now, that we do and it happens to be that we use the mock confog as part of the tagger mock https://github.com/DataDog/datadog-agent/blob/32e9c48ca650d663fc94c311167697cc91e256ab/comp/core/tagger/taggerimpl/tagger_mock.go#L70. 

The config mock, set the global `pkgconfigsetup.Datadog()` as part of the initialization as part of the initialization process https://github.com/DataDog/datadog-agent/blob/c260596a4c871702ab626fda46adb318e140b780/pkg/config/mock/mock.go#L52

If we do not change the test to use the `mockConfig := configmock.New(t)` the code inside that `NewBufferedAggregator` that fetches the global config would not use the right config object.

**Review by sblumenthal — APPROVED** [review]

**Review by dustmop — COMMENTED** [review]
Awesome to see another global being turned into a component!

I have a few style / naming suggestions

**Review by hush-hush — APPROVED** [review]

File: pkg/aggregator/aggregator_test.go (hunk 4 above)
Comment (vickenty) [reply pkg/aggregator/aggregator_test.go:585]: So both the mock module and this code create and install a global config. And this one wins, because it happens last. I guess this works, but is rather unexpected. Would it be possible to have one mock config in the test?

**Review by hmahmood — APPROVED** [review]

File: pkg/aggregator/demultiplexer_agent_test.go [hunk 13]
```diff
@@ -153,15 +156,19 @@ type DemultiplexerAgentTestDeps struct {
...
+	taggerComponent := fxutil.Test[tagger.Mock](t, taggerimpl.MockModule())
+
 	return fxutil.Test[DemultiplexerAgentTestDeps](
 		t,
 		defaultforwarder.MockModule(),
 		core.MockBundle(),
 		orchestratorimpl.MockModule(),
 		eventplatformimpl.MockModule(),
 		compressionimpl.MockModule(),
+		fx.Provide(func() tagger.Component { return taggerComponent }),
```
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply pkg/aggregator/demultiplexer_agent_test.go:172]: Unfortunately the `MockModule` internally uses other mocks, such as config mock module that is also used by the `core.MockModule` so fx complains because we use the same module twice. That is why I had to create the tagger outside

File: pkg/aggregator/no_aggregation_stream_worker.go [hunk 14]
```diff
@@ -199,7 +199,7 @@ func (w *noAggregationStreamWorker) run() {
 							}
 
 							// enrich metric sample tags
-							sample.GetTags(w.taggerBuffer, w.metricBuffer, tagger.EnrichTags)
+							sample.GetTags(w.taggerBuffer, w.metricBuffer, nooptagger.NewTaggerClient().EnrichTags)
```
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply pkg/aggregator/no_aggregation_stream_worker.go:202]: This test was not using the global tagger before, so to keep the test identical I decided to use the nooptagger.

File: pkg/collector/corechecks/cluster/orchestrator/processors/processor.go [hunk 15]
```diff
@@ -12,6 +12,7 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 
 	model "github.com/DataDog/agent-payload/v5/process"
+
```
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply pkg/collector/corechecks/cluster/orchestrator/processors/processor.go:15]: That lines was added by `gofmt` 😄

**Review by vickenty — APPROVED** [review]

File: pkg/aggregator/no_aggregation_stream_worker.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (GustavoCaso) [reply pkg/aggregator/no_aggregation_stream_worker.go:202]: Here is the PR https://github.com/DataDog/datadog-agent/pull/30608
